routing/routing_tests/index_graph_tools.cpp

```diff
@@ -56,12 +59,61 @@ void TestRouteGeometry(IndexGraphStarter & starter,
                        AStarAlgorithm<IndexGraphStarter>::Result expectedRouteResult,
                        vector<m2::PointD> const & expectedRouteGeom)
 {
-  vector<Segment> route;
-  auto const resultCode = CalculateRoute(starter, route);
+  vector<Segment> routeSegs;
+  double timeSec = 0.0;
+  auto const resultCode = CalculateRoute(starter, routeSegs, timeSec);
+
+  if (AStarAlgorithm<IndexGraphStarter>::Result::NoPath == expectedRouteResult
+      && expectedRouteGeom == vector<m2::PointD>())
+  {
+    // The route goes through a restriction. So there's no choice for building route
+    // except for going through restriction. So no path.
+    return;
+  }
+
+  TEST_EQUAL(resultCode, expectedRouteResult, ());
+  if (resultCode != AStarAlgorithm<IndexGraphStarter>::Result::OK)
+    return;
+
+  CHECK(!routeSegs.empty(), ());
+  vector<m2::PointD> geom;
+  for (size_t i = 0; i < routeSegs.size(); ++i)
+  {
+    m2::PointD const & pnt = starter.GetPoint(routeSegs[i], false /* front */);
+    // Note. In case of A* router all internal points of route are duplicated.
+    // So it's necessary to exclude the duplicates.
+    if (geom.empty() || geom.back() != pnt)
+      geom.push_back(pnt);
+  }
+  // @TODO Get rid of the copypase below and get rid of TestRouteSegments.
+  m2::PointD const & pnt = starter.GetPoint(routeSegs.back(), false /* front */);
+  if (geom.empty() || geom.back() != pnt)
```

Comment:
Please, create a lambda, say:
```c++
auto pushPoint = [&geom](m2::PointD const & point) {
  if (geom.empty() || geom.back() != point)
    geom.push_back(point);
};
```
and use this intsead of copy-paste.

FYI: a good practice is to create a Jira task for each individual TODO, and mention task's id here. This increases probability that the issue fill be fixed soon.